### PR TITLE
Support for options with optional arguments 

### PR DESCRIPTION
**Pull request by nikololiahim**
An attempt to solve https://github.com/bkirwi/decline/issues/350. 

For now, it is unclear how an option with optional arguments should be parsed in presence of other options.
For long options, I suggest the following strategy:
* If the next arguments is special, e.g. an option or `--`, this argument should not be consumed and the option with optional argument should be considered empty. 
* Otherwise, the next agument should be consumed.

As for short options, the behavior should be the following (from what I understood of the [POSIX Utitility specification, section 2.b](https://pubs.opengroup.org/onlinepubs/9699919799/)):
* if the option with an optional option-argument is the last in the string of flags, it should be consumed as empty.
* Otherwise, this option should consume the entire remaining string of flags as its option argument.

@bkirwi what do you think?

### CODE REVIEW

File: core/shared/src/main/scala/com/monovore/decline/opts.scala [hunk 1]
```diff
@@ -133,6 +134,33 @@ object Opts {
...
+        case Some(value) => arg.read(value).map(Some(_))
+      }
+
+  def optionalOptArgs[A](
+      long: String,
+      help: String,
+      short: String = "",
+      metavar: String = "",
+      visibility: Visibility = Visibility.Normal
+  )(implicit arg: Argument[A]): Opts[(Int, List[A])] =
```
Comment (bkirwi) [inline core/shared/src/main/scala/com/monovore/decline/opts.scala:156]: What do you think of `NonEmptyList[Option[A]]` instead? (Preserves more ordering information, and it's a "tighter" type in the sense that users don't have to stress about invalid values like `(0, Nil)`.)

File: core/shared/src/main/scala/com/monovore/decline/opts.scala [hunk 2]
```diff
@@ -133,6 +134,33 @@ object Opts {
     Single(Opt.Regular(namesFor(long, short), metavarFor[A](metavar), help, visibility))
       .mapValidated(Argument[A].read)
 
+  def optionalOptArg[A](
```
Comment (bkirwi) [inline core/shared/src/main/scala/com/monovore/decline/opts.scala:137]: Definitely a nitpick, but I'm not sure about the name. (Mix of full words and abbreviations, some of which aren't used elsewhere.) I'm not sure what else I'd suggest, though: maybe `optionalOption`, or `optionFlag` or `flagOption` to suggest you're getting both option-like and flag-like behaviours?

**Review by bkirwi — COMMENTED** [review]
Hey, thank you for the contribution! Some comments inline.

> For long options, I suggest the following strategy:

From testing a few commands I have locally that use optional arguments, like `ls --color` or some git subcommands, it seems like they always expect the argument to be provided with the `=` syntax. (As in `ls --color=always`.) I think I prefer that approach; aside from my strong bias towards "doing the standard thing", it seems like it avoids some ambiguity, and is basically consistent with the short option behaviour.

> As for short options, the behavior should be the following (from what I understood of the [POSIX Utitility specification, section 2.b](https://pubs.opengroup.org/onlinepubs/9699919799/)):

This sounds exactly right. (And thanks for pulling up the spec for it!)

> @bkirwi what do you think?

Overall looks really good! A couple minor suggestions inline.

The last important thing we haven't discussed is updating the usage text. It seems like we'll probably just be able to format this as `--$name[=<$metavar>]` and otherwise treat it like any other option or flag, but I haven't gone through all the details... let me know if you run into any trouble there.

**Comment by nikololiahim** [comment]
Thank you for the quick reply! I more or less understand what to do now.

**Comment by nikololiahim** [comment]
@bkirwi I've implemented the things we discussed + added some tests to ParseSpec and HelpSpec. What do you think?

**Comment by nikololiahim** [comment]
Nevermind, forgot to update the Usage part.

**Comment by bkirwi** [comment]
Great!

I won't have time to take a proper look until next week, unfortunately, but at a first glance things are looking good. (Thanks in particular for updating the tests!)

**Comment by nikololiahim** [comment]
That's great to hear! I'm pretty sure I have missed some important edge case, so don't hesitate to ping me.

File: core/shared/src/main/scala/com/monovore/decline/Parser.scala [hunk 3]
```diff
@@ -95,7 +104,7 @@ private[decline] case class Parser[+A](command: Command[A])
         accumulator
           .parseSub(arg)
           .map { result =>
-            result(rest).swap.map { _.withPrefix(List(command.name)) }.swap.flatMap(evalResult)
+            result(rest).leftMap { _.withPrefix(List(command.name)) }.flatMap(evalResult)
```
Comment (bkirwi) [inline core/shared/src/main/scala/com/monovore/decline/Parser.scala:107]: Good call... I think this dates back to an older Scala version where this API was unavailable.

File: core/shared/src/test/scala/com/monovore/decline/HelpSpec.scala [hunk 4]
```diff
@@ -39,6 +53,10 @@ class HelpSpec extends AnyWordSpec with Matchers {
           |        Second option.
           |    --third <integer>
           |        Third option.
+          |    --flagOpt[=<string>]
```
Comment (bkirwi) [inline core/shared/src/test/scala/com/monovore/decline/HelpSpec.scala:56]: Let's give one of these two new options a short form, to see how that renders here.

(Looking at the code, I suspect we'll get `-F[=<metavar>]`, where the correct rendering would be `-F[<metavar>]`, in which case there's a minor bug to fix. But regardless the test will be useful to prevent any future trouble!)

**Review by bkirwi — COMMENTED** [review]
Very cool! One possible bug, but otherwise this seems thorough and sound.

File: core/shared/src/test/scala/com/monovore/decline/ParseSpec.scala [hunk 5]
```diff
@@ -86,6 +86,82 @@ class ParseSpec extends AnyWordSpec with Matchers with Checkers {
...
+        .orNone
+
+      val all =
+        Opts
+          .flag(long = "all", help = "...", short = "a")
+          .orFalse
+
+      val repeatedFlagOpt = Opts
+        .flagOptions[String](long = "flag", help = "...", short = "f")
+        .map(_.toList)
```
[the remark this answers is not in the document]
Comment (nikololiahim) [reply core/shared/src/test/scala/com/monovore/decline/ParseSpec.scala:101]: True, will update.

File: core/shared/src/test/scala/com/monovore/decline/HelpSpec.scala (hunk 4 above)
Comment (nikololiahim) [reply core/shared/src/test/scala/com/monovore/decline/HelpSpec.scala:56]: Hmm, let me check...

Comment (nikololiahim) [reply core/shared/src/test/scala/com/monovore/decline/HelpSpec.scala:56]: I think it would also make sense to extend the UsageSpec.

Comment (nikololiahim) [reply core/shared/src/test/scala/com/monovore/decline/HelpSpec.scala]: Just checked, you were right. Updated the implementation and tests in https://github.com/bkirwi/decline/pull/415/commits/fd75004574383159275a8a2179a05e7564fd94a2.

File: core/shared/src/test/scala/com/monovore/decline/ParseSpec.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (nikololiahim) [reply core/shared/src/test/scala/com/monovore/decline/ParseSpec.scala]: Updated in https://github.com/bkirwi/decline/pull/415/commits/3131ab349c5425d9af8536b3b8e543ebb750495e.

**Comment by bkirwi** [comment]
The updates look good!

The last thing I need to do is check this for backwards compat. I _think_ this is compatible, in which case I can cut a new minor release with the change. If it's not, it might take slightly longer to merge... but I'll try and get it out soon in any case.
